ctp2_code/ai/strategy/scheduler/scheduler.cpp

```diff
@@ -867,6 +883,231 @@ void Scheduler::Match_Resources(const bool move_armies)
 		}
 	}
 
+	if (needToGoBackOne)
```

Comment:
Wow, did you add all this code?  I am not familiar with the AI so I cannot review this.

In case it is a copy, please try not to copy code but use methods if possible.

Reply:
This is copied from the loop above. The problem this changeset is trying to solve is that the code can remove goals from the list, but would crash if the first goal was removed as it would try to set the iterator to the previous goal, then remove the goal - which doesn't work if the goal to be removed is the first goal.

The fix as implemented catches this case and instead moves the iterator to the next goal and removes the previous one. The loop then moves the iterator on to the item following the one that needs to be processed, so the variable added is used tp . The problem is if there are only 1 or 2 goals. If there is only one goal, this cannot be removed in the iterator loop, otherwise the loop will then be working with an iterator that points to nothing - causing a crash. If there are 2 goals then the loop will terminate without processing the second goal. I therefore added the variable check for this case and repeat the loop to either remove the only item in the list or process it as necessary.

There should be a better way to do this that is overall more elegant and doesn't require copying the loop code (which you are right is a bad idea - it means that any code fixes now need to be made in two places). I contemplate this further.